### PR TITLE
Grating stim class implementation

### PR DESCRIPTION
**Pull request by lightest**
Introduces `GratingStim.js` which helps to create all kinds of grating stimuli. In this particular PR a Gabor patch is introduced.

How it works:
If `tex` parameter passed in is an image name (which was previously loaded using `psychoJS.serverManager`), the stim creates `PIXI.TilingSprite` for it and displays it.
If `tex` is one of the predefined functions the shader is going to be used that will render necessary function.
Same goes for `mask` parameter.

usage:
```JS
import { core } from './lib/psychojs.js';
const { PsychoJS } = core;
const psychoJSInst = new PsychoJS();

psychoJSInst.openWindow({
  fullscr: false,
  color: new util.Color([0, 0, 0]),
  units: 'height',
  waitBlanking: true
});

...

// further, within experiment initialization function
gratingStimInst = new visual.GratingStim({
    win : psychoJSInst.window,
    name: 'gabor',
    tex: 'sin',
    mask: 'gauss',
    spatialFrequency: 3,
    phase: Math.PI
});
```

Questions to address:
~~- How to set size of the patch? The size parameter that could be passed in is used internally in other stimuli like `ImageStim.js` in a fashion normalized coordinates would i.e. it expects [0,1] range there. Introduce another parameter?~~
- How to include shaders in the code? Locally I used esbuild-plugin-glsl which simply enables importing contents of shaders like text files. But the problem here is that `buildSync` does not allows for plugins.

TODO:
~~- Make mask scalable to entire sprite in case requested stim size is greater than provided mask image.~~

### CODE REVIEW

File: src/visual/GratingStim.js [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+ * @param {number} [options.depth= 0] - the depth (i.e. the z order)
+ * @param {number} [options.texRes= 128] - the resolution of the text
+ * @param {boolean} [options.interpolate= false] - whether or not the image is interpolated
+ * @param {boolean} [options.autoDraw= false] - whether or not the stimulus should be automatically drawn on every frame flip
+ * @param {boolean} [options.autoLog= false] - whether or not to log
+ */
+
+export class GratingStim extends util.mix(VisualStim).with(ColorMixin)
+{
+	// win,
```
Comment (lightest) [inline src/visual/GratingStim.js:63]: temporary list of parameters copied from psychoPy as a reference. Do we need to implement all of these eventually?

Comment (peircej) [reply src/visual/GratingStim.js:63]: That would be great, yes

**Comment by peircej** [comment]
What's the grating you've used in the screenshot? It looks like a blurred square wave rather than a sinusoid?

**Review by apitiot — COMMENTED** [review]
An exciting addition to the visual stimuli!

Here are my comments:

**Feature**
- Make sure that the default value for each parameter matches the PsychoPy ones: e.g. "sin" for tex
- Instead of a top-level DEFINED_FUNCTIONS, I would use a "private" member `_shaderMap = new Map<string, string>` defined in the constructor, since it is very much a map associating a tex name or mask name to an actual shader.

**Style**
- I personally do not favour a final or starting dot for numbers: it's a little confusing. Since space is not at a premium, I would go for `10.0`, `0.2`, etc.
- I like to use the suffix `_px` to make explicit the fact that a variable's units is pixel, when there is a possible confusion.
- Make sure to use double quotes instead of single quotes (I'm guilty of that myself).

**Documentation**
- Document all functions and constants, e.g. `_getPixiMeshFromPredefinedShaders`, `DEFAULT_STIM_SIZE`, `DEFINED_FUNCTIONS`
- Make sure that the documentation is complete, e.g. `@return` for all methods (except for void), and correct: e.g. the documentation of `setTex` is still that of `ImageStim.setImage`

**Examples and/or tests**
- We would really need a set of examples, or, at the very least, an example with many different configurations. apitiot/gabor does not make it possible to get an idea of what works and what does not

**Comment by apitiot** [comment]
Thank you. We're almost there, I believe. Just two points:

(a) Could you provide better documentation for the shaders?
They would need a standard header just like the .js files, if only to identify you as the author: credit where credit is due :-)
The text should indicate that this is a shader to be used for GratingStim, and we would need information about what the shader does (not just a wikipedia reference, even though it is indeed good to have one).

(b) #DEFINED_FUNCTIONS
I agree that #DEFINED_FUNCTIONS should be a static member. However I would rename it _Shaders, to be more explicit about what it is.
I understand that a convention for private members is to use # but it seems very confusing to me (jsdoc uses # in a very specific way, # is associated with comments in the mind of most developers), and we are using _ as a prefix for private methods throughout the code so I would suggest we stick to it.
I like very much that default values are defined there: very clean :-)

**Pull request merged**